### ISSUE TITLE
Pattern match and where clause on label not working

### ISSUE THREAD

**Issue by databu** [neo4j/neo4j#5991]
When I run a query with a label constraint either as a pattern match or as a where clause, it neo4j (2.3.1) returns nodes that don't have that label. See attached screenshots.

![neo4j 2015-12-01 12-20-46](https://cloud.githubusercontent.com/assets/1074007/11496934/8bf18f0c-9826-11e5-8f0d-002e0b34d146.png)
![neo4j 2015-12-01 12-21-44](https://cloud.githubusercontent.com/assets/1074007/11496936/8c5323d4-9826-11e5-81ab-0a77c39e4f0b.png)

**Comment by jexp** [comment]
Did you upgrade that store from an earlier version?
Would you be able to share the database privately with us for narrowing down the issue?

**Comment by databu** [comment]
Yes it was upgraded from 2.2.x (I don't remember the exact x). I wouldn't be comfortable sharing the raw db because it has sensitive private data. I could try copying it, removing or anonymising the sensitive bits and see if the problem is still there. It would take me a few days though.

**Comment by jexp** [comment]
Can you (on a backup) delete the schema and index folders in your db?

The information will be rebuild on startup.

You see when the indexes are online with `:schema`

and then try again.

**Comment by databu** [comment]
I removed sensitive data and most unaffected nodes from the db and the problem is still there. This was actually quicker and easier than expected. However I'd still only like to share privately. Please tell me where I should send it.

**Comment by databu** [comment]
Some queries on the reduced DB that I mentioned in my last comment:

match (c) return count(distinct(c)); 
=> 1320

match (c:client) return count(distinct(c));
=> 0

match (c:driver) return count(distinct(c));
=> 245

But:
match (c:driver) where 'client' in labels(c) return count(distinct(c));
=> 245

And:
match (c) where not('client' in labels(c)) and not('driver' in labels(c)) return count(distinct(c));
=> 0

**Comment by databu** [comment]
jexp , sorry I had overlooked your latest comment. I deleted schema and index folders and it works ok now:

match (c) return count(distinct(c)); 
=> 1320

match (c:client) return count(distinct(c));
=> 1320

match (c:driver) return count(distinct(c));
=> 0

But:
match (c:driver) where 'client' in labels(c) return count(distinct(c));
=> 0

And:
match (c) where not('client' in labels(c)) and not('driver' in labels(c)) return count(distinct(c));
=> 0

Thanks. However I could still share the erroneous DB if it helps to prevent this from happening again.
